k2/csrc/fsa_utils.cu

```diff
@@ -603,45 +604,35 @@ Ragged<int32_t> GetStateBatches(FsaVec &fsas, bool transpose) {
   // it passes the end of this FSA, we get the beginnings of the batches
   // we want.  The natural algorithm to find the beginnings of the batches
   // is sequential.
-  Array2<int32_t> dest_states_powers(c, num_states, log_power + 1);
+  Array2<int32_t> dest_states_powers(c, log_power + 1, num_states);
   const int32_t *arc_dest_states_data = arc_dest_states.Data(),
                 *fsas_row_splits2_data = fsas.RowSplits(2).Data();
-  int32_t *dest_states_data = dest_states_powers.Data();
+  int32_t *dest_states_power_data =
+      dest_states_powers.Data();  // only process Row[0] below
   const int32_t int_max = std::numeric_limits<int32_t>::max();
   auto lambda_set_dest_states =
       [=] __host__ __device__(int32_t state_idx01) -> void {
     int32_t arc_idx01x = fsas_row_splits2_data[state_idx01],
-            next_arc_idx01x = fsas_row_splits2_data[arc_idx01x];
+            next_arc_idx01x = fsas_row_splits2_data[state_idx01 + 1];
     // If this state has arcs, let its `dest_state` be the largest `dest_state`
     // of any of its arcs (which is the last one); otherwise, take the
     // `dest_state` from the 1st arc of the next state, which is the largest
     // value we can take (if the definition is: the highest-numbered state s for
     // which neither this state nor any later-numbered state has an arc to a
     // state lower than s).
-    int32_t arc_idx012 = max(arc_idx01x, next_arc_idx01x - 1);
+    int32_t arc_idx012 =
+        max(arc_idx01x,
+            next_arc_idx01x - 1);  // if this state has arcs, next_arc_idx01x -
```

Comment:
Does this assume that arcs of a state are sorted by its dest_state ?
The comment of this function only  says that the input fsas is topologically sorted.

Reply:
No, we didn't assume it's sorted by dest_state. We just get the last arc index  in this state (well, if it has arcs). 
Then `arc_dest_states_data[arc_idx012]` will give the next state we should go forward (in the chain). Notice that we get lower bound of original dest_state to get `arc_dest_states`.